crypto/ml_kem/ml_kem.c

```diff
@@ -0,0 +1,1567 @@
+        out->c[i + 1] = reduce_once(value);
+    }
+    return 1;
+}
+
+/*
+ * Algorithm 7 from the spec, with eta fixed to three and the PRF call
+ * included. Creates binominally distributed elements by sampling 3*|eta| bits,
+ * and setting the coefficient to the count of the first bits minus the count of
+ * the second bits, resulting in a centered binomial distribution.
+ */
+int ossl_ml_kem_cbd_3(scalar *out,
+                      uint8_t in[ML_KEM_RANDOM_BYTES + 1],
+                      mctx *ctx)
+{
+    uint8_t randbuf[6 * DEGREE / 8];    /* 64 * eta */
+    int i = 0, j = 0;
+    uint8_t b1, b2, b3;
+    uint16_t value;
+
+# define bit0(b) (b & 1)
+# define bitn(n, b) ((b >> n) & 1)
+
+    if (!prf(randbuf, sizeof(randbuf), in, ctx))
+        return 0;
+    /* Unrolled loop uses 3 bytes at a time, yielding 4 values (6 bits each) */
+    while (j < (int) sizeof(randbuf)) {
+        b1 = randbuf[j++];
+        b2 = randbuf[j++];
+        b3 = randbuf[j++];
+
+        value = kPrime + bit0(b1) + bitn(1, b1) + bitn(2, b1);
+        value -= bitn(3, b1)  + bitn(4, b1) + bitn(5, b1);
+        out->c[i++] = reduce_once(value);
+
+        value = kPrime + bitn(6, b1) + bitn(7, b1) + bit0(b2);
+        value -= bitn(1, b2) + bitn(2, b2) + bitn(3, b2);
+        out->c[i++] = reduce_once(value);
+
+        value = kPrime + bitn(4, b2) + bitn(5, b2) + bitn(6, b2);
+        value -= bitn(7, b2) + bit0(b3) + bitn(1, b3);
+        out->c[i++] = reduce_once(value);
+
+        value = kPrime + bitn(2, b3) + bitn(3, b3) + bitn(4, b3);
+        value -= bitn(5, b3) + bitn(6, b3) + bitn(7, b3);
+        out->c[i++] = reduce_once(value);
+    }
+# undef bit0
+# undef bitn
+
+    return 1;
+}
+
+void ossl_ml_kem_encode_public_key(uint8_t *out,
+                                   const scalar *t,
+                                   const uint8_t rho[ML_KEM_RANDOM_BYTES],
+                                   vinfo_t vinfo)
+{
+    vector_encode_12(out, t, vinfo->rank);
+    memcpy(out + vinfo->vector_bytes, rho, ML_KEM_RANDOM_BYTES);
+}
+
+int ossl_ml_kem_parse_public_key(const uint8_t *in,
+                                 scalar *m, scalar *t,
+                                 uint8_t rho[ML_KEM_RANDOM_BYTES],
+                                 uint8_t pkhash[ML_KEM_PKHASH_BYTES],
+                                 vinfo_t vinfo,
+                                 mctx *ctx)
+{
+    /* Decode and check |t| */
+    if (!vector_decode_12(t, in, vinfo->rank))
+        return 0;
+    /* Save the matrix |m| recovery seed |rho| */
+    memcpy(rho, in + vinfo->vector_bytes, ML_KEM_RANDOM_BYTES);
+    /*
+     * Pre-compute the public key hash, needed for both encap and decap.
+     * Also pre-compute the matrix expansion, stored with the public key.
+     */
+    return hash_h(pkhash, in, vinfo->pubkey_bytes, ctx)
+        && matrix_expand(m, rho, vinfo->rank, ctx);
+}
+
+void ossl_ml_kem_encode_private_key(uint8_t *out,
```

Comment:
This seems to be 'dk' from Algorithm 16.

Reply:
Yes, it is. I think you're asking for a comment noting this. I'll add one.